source/log.c

```diff
@@ -0,0 +1,386 @@
+static aws_log_report_fn* s_log_report_callback = s_aws_log_default_report_function;
+
+AWS_THREAD_LOCAL struct aws_log_context *s_local_log_context;
+
+void aws_log_set_reporting_callback(aws_log_report_fn* report_callback) {
+    if (report_callback) {
+        s_log_report_callback = report_callback;
+    } else {
+        s_log_report_callback = s_aws_log_default_report_function;
+    }
+}
+
+int aws_log_system_init(struct aws_allocator *alloc, size_t max_message_len, int memory_pool_message_count, int max_message_count) {
+    if (max_message_len < 1 || max_message_count < 1) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return AWS_OP_ERR;
+    }
+
+    int zero = 0;
+    if (!aws_atomic_compare_exchange(&s_log_system_running, &zero, 1)) {
+        aws_raise_error(AWS_ERROR_LOG_DOUBLE_INITIALIZE);
+        return AWS_OP_ERR;
+    }
+
+    s_log_system_params.max_message_len = max_message_len;
+    s_log_system_params.memory_pool_message_count = memory_pool_message_count;
+    s_log_system_params.max_message_count = max_message_count;
+    s_log_system_params.alloc = alloc;
+
+    aws_log_thread_init();
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_log_system_clean_up() {
+    s_log_system_running = 0;
+    aws_log_thread_clean_up();
+    aws_log_flush();
+}
+
+int aws_vlog(enum aws_log_level level, const char *fmt, va_list va_args) {
+    if (!s_local_log_context) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    if (aws_atomic_add(&s_log_message_count, 1) >= s_log_system_params.max_message_count) {
+        aws_raise_error(AWS_ERROR_LOG_THREAD_MAX_CAPACITY);
+        aws_atomic_add(&s_log_message_count, -1);
+        return AWS_OP_ERR;
+    }
+
+    /* Atomically grab and release all messages on the `delete_list` whenever present. */
+    struct aws_log_message *delete_list;
+    do {
+        delete_list = s_local_log_context->delete_list;
+    } while (!aws_atomic_compare_exchange_ptr((void **)&s_local_log_context->delete_list, (void **)&delete_list, NULL));
+
+    while (delete_list) {
+        struct aws_log_message *next = delete_list->next;
+        aws_memory_pool_release(s_local_log_context->message_pool, delete_list);
+        delete_list = next;
+    }
+
+    /* Acquire memory for the new message. */
+    void *mem = aws_memory_pool_acquire(s_local_log_context->message_pool);
+    if (!mem) {
+        return AWS_OP_ERR;
+    }
+    struct aws_log_message *msg = (struct aws_log_message *)mem;
+    char *msg_data = (char*)(msg + 1);
+    
+    /* Format the message. */
+    char date[256];
+    time_t now = time(NULL);
+#ifdef _MSC_VER
+    struct tm *t = localtime(&now);
+    strftime(date, sizeof(date) - 1, "%m-%d-%Y %H:%M:%S:%Z", t);
+#else
+    struct tm t;
+    localtime_r(&now, &t);
+    strftime(date, sizeof(date) - 1, "%m-%d-%Y %H:%M:%S:%Z", &t);
+#endif
+
+    char fmt_final[1024];

```

Comment:
I don't quite understand how it could make sense to format into the same buffer that is to be read from? At first glance it sounds like `snprintf` could overwrite data before reading it in many cases.